src/shogun/base/AnyParameter.h

```diff
@@ -28,101 +28,85 @@ namespace shogun
 		GRADIENT_AVAILABLE = 1
 	};
 	/** @brief Class AnyParameterProperties keeps track of parameter properties.
-	 *
-	 * Currently there are three types of parameters:
-	 * -# HYPERPARAMETER: the parameter determines how the training is
-	 * performed, e.g. regularisation
-	 * -# GRADIENT_PARAM: the parameter is used in gradient updates
-	 * -# MODEL_PARAM: model parameter that is trained, e.g. weight and bias
-	 * vectors and bias
+	 * The parameter properties can be either true or false.
+	 * These properties describe if a parameter is for example a hyperparameter
+	 * or if it has a gradient.
 	 */
 	class AnyParameterProperties
 	{
 	public:
-		static const int32_t HYPERPARAMETER = 0x00000001;
-		static const int32_t GRADIENT_PARAM = 0x00000010;
-		static const int32_t MODEL_PARAM = 0x00000100;
+		static const int32_t HYPER = 1u << 0;
+		static const int32_t GRADIENT = 1u << 1;
+		static const int32_t MODEL = 1u << 2;
```

Comment:
ah wait. the thing I had in mind was 
```
static const int HYPER = 1;
```
and then
```
mask = (1 << HYPER) | (1 << MODEL);
```
This is now difference, as you are effectively storing masks themselves in these const members, not the number of bit shifts. Or am I wrong? Let me stare a bit more to understand

Reply:
Do you mean:
```
static const int HYPER = 0;
static const int GRADIENT = 1;
static const int MODEL = 2;
```

Reply:
yes, but start at 1

Reply:
OK, if we have:
```
  static const int HYPER = 1;
  static const int GRADIENT = 2;
  static const int MODEL = 3;

  bool model_selection = true;
  bool gradient = true;

  int mask = (1 << HYPER) & model_selection | (1 << GRADIENT) & gradient;
```
Then the above mask will always be false/0 no?

Reply:
Do you mean more like this:
```
int mask = (model_selection << HYPER) | (gradient << GRADIENT)
```
which is the equivalent of shift bit at position X (1,2 or 3 in this case) if the function parameter is true.

Reply:
sure, the thing you wrote is (or should be, not sure what happens when you shift a boolean true) a neater version if what I wrote above.

Reply:
Ah now I remember why I did it this way.. Basically if you want to provide an interface to create your own mask like this
```
AnyParameterProperties("My description",  
                                           AnyParameterProperties::MODEL |
                                           AnyParameterProperties::GRADIENT)
```
you need to have the values of the static member with the shift. Unless I am missing something?

Reply:
Not sure what you mean, but what I have seen (and been doing) would result in
```
AnyParameterProperties("My description",  
                       (1 << AnyParameterProperties::MODEL) |
                       (1 << AnyParameterProperties::GRADIENT))
```
It is not as neat, you are totally right!. BUT it can be hidden quite easily either.
I mean, it is your code, Ill leave it to you. Not sure whether it is possible to say what is better now, we will have to commit to something and then see if it works :) So go ahead with the above solution and then we see where we get

Reply:
OK! Btw why would does it need to start at 1, instead of 0?

Reply:
ah, no that was nonsense (just needed to make sure that you never get an all zero mask)

BTW I also realised why the micro-controller people store the number of bits rather than the full masks: it uses less memory (which is not really relevant here, but still good to think about it for a moment).
```
byte HYPER = 2
...
byte BLA = 255
```
so with an 8bit integer, you can do 255 attributes. If you wanted to store the mask for each attribute instead (each mask is 8bit), you would have to store 255 * 8bit, which is quite a difference , especially if you have only 4k memory as on an AVR ;)

How many parameters are registered during a shogun run? Let's say 100 is realistic. Then we have 100 byte usage if we store only the offset, and 2.5kB if we use the masks. In a data heavy environment, and since we won't register many many more parameters this doesn't matter. For 1000 parameters it is still ok, 10^6 will result in some pretty big differences.

Reply:
Lets typedef the actual type so we can change later.